### ISSUE TITLE
Alias of Final at import no allowed

### ISSUE THREAD

**Issue by Yobmod** [microsoft/pyright#966]
I usually import longer types as shorter aliases , but pylance/pyright complains about this for Final. It does not complain about other types like Literal as Lit or Optional as Opt). Should this be allowed? (mypy doesnt complain about it).

python 3.7, VS-code 1.48.1, pylance v2020.8.2

```
from typing_extensions import Final as Fin, Literal as Lit

a: Fin[int] = 100   # "Final" is not allowed in this context
b: Lit['b'] = 'b'      # no error
```

```
from typing_extensions import Final

a: Final[int] = 100   # no error
```

**Comment by erictraut** [comment]
One of the reasons Pyright is so much faster than mypy and other type checkers is that it uses a technique called "lazy type evaluation". When it's evaluating the type of an expression or a symbol, it performs the minimum work necessary. For example, if an expression references a function that is imported from another module, it will evaluate only the function signature within that imported module, not the entire imported module — and not even the body of that function (unless it needs to infer the return type). Mypy, by contrast, is a multi-pass analyzer that does analysis of entire files in a top-down manner.

To perform lazy analysis, Pyright performs a fast initial step called "binding". During the binding phase, the symbols for each scope are identified, but their types are not yet known. (This may be more information than you care about, but if you want additional details, check out [this documentation](https://github.com/microsoft/pyright/blob/master/docs/internals.md) ).

Lazy type analysis is challenging in a language like Python because of the paucity of keywords. In most languages, there would be keywords for type aliases and finals. In Python, these are specified using non-keyword types defined in the typing module. There are three such identifiers that need to be understood during or before the binding phase in Pyright: "Final", "Literal", and "TypeAlias". (TypeAlias will appear in Python 3.10 as defined in PEP 613.) For these three symbols, Pyright uses a bit of a hack because it can't evaluate their types at binding time. As a workaround, Pyright looks specifically for expression of the form "Final", "typing.Final", and "t.Final" and assumes that these will evaluate to the type "Final" in the typing module. If you use an expression other than those three (such as "Fin"), this workaround fails.

I'll take another look at whether there are more robust ways to handle this, but I suspect this may be a limitation that we live with.

In any case, I would encourage you _not_ to use aliases like this in your code for standard typing constructs. You are saving almost nothing in terms of screen real estate and keystrokes, and it makes your code hard to read for anyone else who is not familiar with your approach. If anyone on my team attempted to submit a PR with code that did this, that PR would be promptly rejected! :)

**Comment by erictraut** [comment]
I found a way to make this more robust. There are still extreme edge cases where it will fail, but all common cases such as your reported use case will now work as expected. This will be included in the next release of Pyright and Pylance.

**Comment by erictraut** [comment]
This is now fixed in Pyright 1.1.65, which I just published. It will also be included in the next version of Pylance.
